### DISCUSSION TITLE
Set bin size to 1 for binX

### DISCUSSION THREAD

**Discussion by mkfreeman** [observablehq/plot#2233] [Q&A] [answered]
Is it possible to specify to `Plot.binX()` that I want only one observation per bin? For context, I want to create a bar chart of the `value` column in my dataset for each `date`. A few tricky pieces:
- Some dates are missing, so I can't simply use `barY`
- I don't want Plot to do any aggregation of the values, I want to see the exact values in the dataset (so _hinting_ at a number of intervals isn't sufficient)
- The time interval in the data may change different (e.g., it may be each `day`, `hour`, etc.) so I can't simply specify the `interval` option

It seems that a `rect` mark with binning is the appropriate method, but I would want to ensure that there was only one observation per bin.

See a few approaches in this notebook: https://observablehq.com/d/4a1a0770b9e533b1

**Comment by Fil** [comment, marked as answer]
Fun. There is no solution if two data points have the same exact date. So you have to make the assumption that they are all different.

Then I suppose that the desired interval can be defined as a heuristic, such as the minimum of the differences between consecutive dates, maybe rounded to account for durations that overlap the  DST:

```js
interval = {
  const m = d3.min(
    d3.pairs(d3.sort(Plot.valueof(fakeData, "date")),
    ([a, b]) => Math.round((b - a) / 60000) // minutes
  );

  return m < 2
    ? d3.utcMinute
    : m < 10
    ? d3.utcMinute.every(m)
    : m < 60
    ? d3.utcMinute.every(m)
    : m < 23 * 60
    ? d3.utcHour.every(Math.round(m / 60))
    : m < 23 * 60 * 7
    ? d3.utcDay
    : m < 24 * 60 * 27
    ? d3.utcWeek
    : m < 24 * 60 * 364
    ? d3.utcMonth
    : d3.utcYear;
}
```

I haven't really tested and it's incomplete (milliseconds, quarters, centuries etc)

**Reply by mkfreeman** [reply]
Thanks @Fil 🫶 !
